### PR TITLE
Fix incorrect aggregation of volume mount and dynamic volumes created via Hera

### PR DESCRIPTION
**Pull request by flaviuvadan**
**Pull Request Checklist**
- [x] Fixes #
- [x] Tests added
- [x] Documentation/examples added
- [x] [Good commit messages](https://cbea.ms/git-commit/) and/or PR title

**Description of PR**
Hera supports building dynamic volumes via the `Volume` object. That creates PVCs + PVs dynamically. Users have the liberty to mount these volumes wherever they want. However, if users have a volume mount _already_, then the dynamic volumes are not combined with mounts correctly. That happens because of the incorrect conditions inside of `build_volume_mounts`. This PR fixes that

### CODE REVIEW

File: src/hera/workflows/_mixins.py [hunk 1]
```diff
@@ -539,25 +539,25 @@ def _build_volume_mounts(self) -> Optional[List[VolumeMount]]:
...
         if result is None and self.volume_mounts is None:
             return None
         elif result is None and self.volume_mounts is not None:
             return self.volume_mounts
+        elif result is not None and self.volume_mounts is None:
+            return result
 
-        mounts = cast(List[VolumeMount], self.volume_mounts) or [] + cast(List[VolumeMount], result) or []
-        return mounts or None
+        return self.volume_mounts + result
```
Comment (elliotgunton) [inline src/hera/workflows/_mixins.py:560]: Think this still needs a cast to pass lint checks

File: src/hera/workflows/_mixins.py [hunk 2]
```diff
@@ -539,25 +539,25 @@ def _build_volume_mounts(self) -> Optional[List[VolumeMount]]:
...
-        else:
+        if self.volumes is None:
             volumes = []
+        else:
+            volumes = self.volumes if isinstance(self.volumes, list) else [self.volumes]
 
         result = (
             None
-            if self.volumes is None
+            if volumes is []
```
Comment (elliotgunton) [inline src/hera/workflows/_mixins.py:549]: Using `is` here is ringing alarm bells as it's only for `None` really (should use `==` or directly check on `volumes`). We could use a `not volumes` or swap the if/else around to use `if volumes`?
```suggestion
            if not volumes
```

**Comment by flaviuvadan** [comment]
@elliotgunton applied your suggestions

**Review by elliotgunton — APPROVED** [review]
:shipit:

**Pull request merged**